### DISCUSSION TITLE
Non-Modal Show, TailwindCSS, and Disabled Items

### DISCUSSION THREAD

**Discussion by elijahmanor** [johnlindquist/kit#78] [Ideas]
![Kapture 2021-03-26 at 12 59 17](https://user-images.githubusercontent.com/86454/112674478-fcc24b00-8e33-11eb-9230-e535c2ed4176.gif)

I've been working on the above Time Log script (see code below) and I've run into a few items that I thought I'd bring up for discussion.

1. I'm currently showing 3 items (Show Logs, Check In, and Check Out). Ideally, I'd rather disable Check Out if you haven't started working yet and disable Check In if you are currently working. I have code inside of those handlers to ignore the action in case it is called when unnecessary, but thought having the item disabled might work better. I **could* just not show the other items at all as a solution.
2. I'm currently displaying the summary information with the `show` method and styling with TailwindCSS classes, however, I ran into several that weren't available so for now I'm hardcoding those with the `style` property.  In particular I was trying to use grid classes and `list-none`.
3. I'd also like a way to show the summary without opening a separate modal. It'd be optimal if I could have a results tab, or show how display the results at the bottom of the kit prompt, or some other solution. It'd be nice if I could either just sent a basic string and it display nicely OR provide HTML formatted with Tailwind for full control. I understand there are CLI implications of this and I'm not sure exactly what should happen there. Maybe the user can conditionally send more than one renderer (one for web and one for terminal using chalk, etc.)
4. Since I mentioned `show`, I noticed that it accepted options, but I couldn't trace down what those options could be. Can you set the height and width? In the GIF above the dialog is much bigger than needed. It'd be nice to suggest a particular size. I also noticed that sometimes I have multiple of these show dialogs open after I've used it a few times. That is my bad for keeping the previous one open, but it seems to happen to me numerous times.

<details>

<summary>Click to Expand Code</summary>

```js
// Menu: Time Log
// Description: Clock In & Clock Out
// Author: Elijah Manor
// Twitter: @elijahmanor

let { isToday, differenceInMinutes, addMinutes } = await npm("date-fns");

const scriptDB = db("time-log", { logs: [] });
let logs = scriptDB.get("logs").value();
let today = logs.find((log) => isToday(new Date(log.date)));

if (!today) {
  today = {
    date: new Date().toUTCString(),
    stamps: [],
  };
  logs.push(today);
}

const timeOptions = {
  hour: "2-digit",
  minute: "2-digit",
  hour12: true,
};

const minutesToHours = (mins) => {
  const negative = mins < 0;
  const hours = Math.abs(mins) / 60;
  const rhours = Math.floor(hours);
  const minutes = (hours - rhours) * 60;
  const rminutes = Math.round(minutes);
  return `${negative ? "-" : " "}${rhours
    .toString()
    .padStart(2, "0")}:${rminutes.toString().padStart(2, "0")}`;
};

const minutesWorkedToday = today.stamps.reduce((memo, item) => {
  const stamp = { ...item };
  if (!stamp.out) {
    stamp.out = new Date().toUTCString();
  }
  memo += differenceInMinutes(new Date(stamp.out), new Date(stamp.in));
  return memo;
}, 0);

const generateHtml = () => {
  const logs = today.stamps.reduce((memo, stamp, index) => {
    memo += `<span class="font-semibold text-green-900">In</span><time class="font-mono text-green-900">${new Date(
      stamp.in
    ).toLocaleTimeString("en-US", timeOptions)}</time>`;
    memo +=
      stamp.out != null
        ? `<span class="font-semibold text-red-900">Out</span><time class="font-mono text-red-900">${new Date(
            stamp.out
          ).toLocaleTimeString("en-US", timeOptions)}</time>`
        : `<span class="font-semibold text-blue-900">Now</span><time class="font-semibold font-mono text-blue-900">${new Date().toLocaleTimeString(
            "en-US",
            timeOptions
          )}</time>`;
    if (index < today.stamps.length - 1) {
      const nextStamp = today.stamps[index + 1];
      const difference = differenceInMinutes(
        new Date(nextStamp.in),
        new Date(stamp.out)
      );
      memo += `<span class="text-gray-900">Break</span><time class="font-mono text-gray-900">${minutesToHours(
        difference
      )} HR</time>`;
    }
    return memo;
  }, "");

  let totals = `<li class="list-none" style="list-style-type: none;"><time class="font-mono">${minutesToHours(
    minutesWorkedToday
  )} HR</time> Total Worked</li>
<li class="list-none" style="list-style-type: none;"><time class="font-mono">${minutesToHours(
    480 - minutesWorkedToday
  )} HR</time> Time Remaining</li>
<li class="list-none" style="list-style-type: none;"><time class="font-mono">${addMinutes(
    new Date(),
    480 - minutesWorkedToday
  ).toLocaleTimeString("en-US", timeOptions)}</time> Estimated End</li>`;
  const summary =
    minutesWorkedToday >= 480
      ? `<div class="mt-4 font-semibold">You've worked <span class="font-mono">8</span> hours today!</div>`
      : "";

  return `<div class="m-4">
    <div class="grid grid-cols-2 gap-4" style="display: grid; grid-template-columns: repeat(2, minmax(0, 1fr)); gap: 0.25rem;">${logs}</div>
    <ul class="mt-4 ml-0 list-none" style="list-style-type: none;">${totals}</ul>
    ${summary}
</div>`;
};

const isAlreadyStarted = today.stamps.some((s) => s.in && !s.out);

let totals = `, , `;

const actions = [
  {
    name: "View Logs",
    value: "View Logs",
    description: `${minutesToHours(
      minutesWorkedToday
    )} HR Total, ${minutesToHours(
      480 - minutesWorkedToday
    )} HR Remaining, ${addMinutes(
      new Date(),
      480 - minutesWorkedToday
    ).toLocaleTimeString("en-US", timeOptions)} Estimated End`,
  },
  {
    name: "Clock In",
    value: "Clock In",
    description: isAlreadyStarted ? `Currently working...` : null,
    // html: isAlreadyStarted ? `<div class="bg-gray">test</div>` : null
  },
  {
    name: "Clock Out",
    value: "Clock Out",
    description: !isAlreadyStarted ? `Currently not working.` : null,
  },
];
const action = await arg("Select an action...", actions);

if (action === "Clock In") {
  if (!isAlreadyStarted) {
    today.stamps.push({ in: new Date().toUTCString() });
  }
  scriptDB.set("logs", logs).write();
} else if (action === "Clock Out") {
  const entry = today.stamps.find((s) => s.in && !s.out);
  if (entry) {
    entry.out = new Date().toUTCString();
  }
  scriptDB.set("logs", logs).write();
} else {
  show(generateHtml());
}
```

</details>

[Install Time-Log Script](https://scriptkit.app/api/new?name=time-log&url=https://gist.githubusercontent.com/elijahmanor/1be041bb238c3205c53e7f7a5df1d7e4/raw/e234169a0ee512abcb4666746fc1f4ec0afdddec/time-log)

**Comment by johnlindquist** [comment]
❤️ the script, thank you so much for sharing!

1. I'll need to think through the scenarios of disabling vs removing the item completely. It seems like most disabled list items are disabled for a reason (usually some state somewhere) and surfacing that state and hiding the items might make more sense. Although there's not a lot of places to display state in the UI at the moment... 🤔

2. I think I just need to include _all_ tailwind classes somehow. I'm currently using a [safelist here](https://github.com/johnlindquist/kitapp/blob/beta/tailwind.config.js#L10) because disabling purging results in a 13MB CSS file. So I need to find a happy medium between "all" and "_everything_" :D

3. I agree 100%. I think I'll create a `panel()` fn where you can pass arbitrary HTML/markdown  (maybe even React/mdx 🤔?). There are terminal libraries that can take React components. [Ink](https://www.npmjs.com/package/ink) looks particularly compelling. The app version will definitely come first, I'm just glad there's a path forward with the terminal.

4. The `show` options ultimately end up [here](https://github.com/johnlindquist/kitapp/blob/beta/src/show.ts#L64). So they are [BrowserWindow options](https://www.electronjs.org/docs/api/browser-window) that will override any of the defaults I set. You should even be able to control things like transparent background which makes for neat tricks like Twitch overlays 😊
